segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/TableBasedMetadataStore.java

```diff
@@ -167,25 +181,32 @@ private StorageMetadataException handleException(Throwable e) throws StorageMeta
         return new StorageMetadataException("Transaction failed", e);
     }
 
-    private void ensureInitialized() {
+    private CompletableFuture<Void> ensureInitialized() {
         if (!isTableInitialized.get()) {
-            try {
-                this.tableStore.createSegment(tableName, timeout).join();
-                log.info("Created table segment {}", tableName);
-            } catch (CompletionException e) {
-                if (e.getCause() instanceof StreamSegmentExistsException) {
-                    log.info("Table segment {} already exists.", tableName);
-                }
-            }
-            isTableInitialized.set(true);
+            return this.tableStore.createSegment(tableName, timeout)
+                    .thenRunAsync(() -> {
+                        log.info("Created table segment {}", tableName);
+                        isTableInitialized.set(true);
+                    }, getExecutor())
+                    .exceptionally(e -> {
+                        val ex = Exceptions.unwrap(e);
+                        if (e.getCause() instanceof StreamSegmentExistsException) {
+                            log.info("Table segment {} already exists.", tableName);
+                            isTableInitialized.set(true);
```

Comment:
I wonder why the initial `if` statement is false. The first thread that gets to create the table will set it to `true` after that all other calls of this method shouldn't get inside the `if` block but we saw otherwise and `Table segment...` message was kept logging even after 2 hours of creation. 

My initial thought was that `isTableInitialized` doesn't have `volatile` keyword but it's of `AtomicBoolean` type so that shouldn't be required. 

Still curious to know why the logic doesn't work without such explicit set to `true` in `exceptionally` block

Reply:
On a freshly installed system there won't exist any storage metadata segment. Once this table segment is created, then from that point onwards we never recreate it - only use existing one. 

When a container is restarted (eg. during failover), then Storage and its TableBasedMetadataStore instances are recreated every time but the persistent  storage metadata segment already exists. 

There is no way for code to know whether it is running for the very first time or after restart - other than check the existence of table segment.  Either way this check needs to happen only once for current incarnation of the container instance.  isTableInitialized tracks that.

StreamSegmentExistsException also means that table is confirmed to be already initialized .